### PR TITLE
Refactor cpp code in gold/digit-dp, add java implementation

### PR DESCRIPTION
**Pull request by ChuangSheep**
Java Impl is coming some time next week

_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
image
LMAOOOOO

File: content/4_Gold/Digit_DP.mdx [hunk 1]
```diff
@@ -92,61 +115,280 @@ void reset() {
...
+				}
+			}
+		}
+	}
+
+	/**
+	 * Calculate the number of sequences whose digits consist of at least half
+	 * targ. If targ2 is not -1, then calculate the number of sequences whose
+	 * digits are exactly half targ and half targ2.
+	 *
```
Comment (SansPapyrus683) [inline content/4_Gold/Digit_DP.mdx:262]:
```suggestion
```

File: content/4_Gold/Digit_DP.mdx [hunk 2]
```diff
@@ -92,61 +115,280 @@ void reset() {
...
+	 * @param pos     the starting position
+	 * @param k       counter for the number of the target digit targ
+	 * @param under   1 if the current sequence is smaller than the given
+	 *     upperbound
+	 * @param started 1 if the sequence has any digit other than leading zeros
+	 */
+	public long solveDP(int pos, int k, int under, int started, int targ,
+	                    int targ2) {
```
Comment (SansPapyrus683) [inline content/4_Gold/Digit_DP.mdx:270]: bro what is this function formatting :sob:

File: content/4_Gold/Digit_DP.mdx [hunk 3]
```diff
@@ -92,61 +115,280 @@ void reset() {
...
+
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("odometer.in"));
+		StringTokenizer st = new StringTokenizer(br.readLine());
+		br.close();
+
+		long X = Long.parseLong(st.nextToken());
+		long Y = Long.parseLong(st.nextToken());
+
+		Odometer solver = new Odometer();
```
Comment (SansPapyrus683) [inline content/4_Gold/Digit_DP.mdx:380]: why make a new class? idt the c++ had to do anything like that...

**Review by SansPapyrus683 — APPROVED** [review]

File: content/4_Gold/Digit_DP.mdx (hunk 2 above)
Comment (ChuangSheep) [reply content/4_Gold/Digit_DP.mdx]: image
sry ci hates it

File: content/4_Gold/Digit_DP.mdx [hunk 4]
```diff
@@ -43,45 +44,67 @@ frequency: 1
...
+involves considering all 9 digits one at a time to occupy at least half of the
+number.
 
-Let $\texttt{dp[i][und][k][is0]}$ where $i$ is the current position, $und$ is whether you have gone below the actual number, $k$ is a counter to see if you have at least half of the same digit, and $is0$ is a boolean if you still have leading zeros. The transition involves looping through 9 digits to add to the current location and comparing with the digit we want to occupy at least half of the number.
+Let $\texttt{dp[pos][k][under][started]}$ where $pos$ is the current position,
+$k$ is a counter to see if you have at least half of the same digit, $under$ is
+whether you have gone below the actual number, and $started$ is a boolean if you
+still have leading zeros. The transition involves looping through 9 digits to
+add to the current location and comparing with the digit we want to occupy at
+least half of the number.
```
Comment (ryanchou-dev) [inline content/4_Gold/Digit_DP.mdx:76]: texttt up the `pos`, `under`, etc.

File: content/4_Gold/Digit_DP.mdx [hunk 5]
```diff
@@ -92,61 +115,276 @@ void reset() {
...
-		if (!started) return 0;
+/**
+ * Calculate the number of sequences whose digits consist of at least half targ.
+ * If targ2 is not -1, then calculate the number of sequences whose digits are
+ * exactly half targ and half targ2.
+ * @param pos the starting position
+ * @param k counter for the number of the target digit targ
+ * @param under if the current sequence is smaller than the given upperbound
+ * @param started if the sequence has any digit other than leading zeros
+ */
```
Comment (ryanchou-dev) [inline content/4_Gold/Digit_DP.mdx:126]: ONGGG

File: content/4_Gold/Digit_DP.mdx [hunk 6]
```diff
@@ -92,61 +115,276 @@ void reset() {
...
+				} else {
+					newK = k - 1;
+				}
+			}
+			ans += solveDP(pos + 1, newK, isUnder, isStarted, targ, targ2);
+		}
+		return dp[pos][k][under][started] = ans;
+	}
+
+	/** Count interesting sequences that are less or equal ubound */
```
Comment (ryanchou-dev) [inline content/4_Gold/Digit_DP.mdx:347]:
```suggestion
	// Count interesting sequences that are less or equal ubound
```

**Review by ryanchou-dev — COMMENTED** [review]
i love you chuyang <3

File: content/4_Gold/Digit_DP.mdx (hunk 6 above)
Comment (ChuangSheep) [reply content/4_Gold/Digit_DP.mdx:347]: emm we do need `**` for function doc don't we?

Comment (SansPapyrus683) [reply content/4_Gold/Digit_DP.mdx:347]: yeah dw about this

**Pull request merged**